File: python/src/reality_capture/specifications/tiling.py
from pydantic import BaseModel, Field, constr
from typing import Optional
from enum import Enum
from geometry import BoundingBox, Point3d


class TilingInputs(BaseModel):
    scene: str = Field(description="Reality data id of ContextScene to process")
    region_of_interest: Optional[str] = Field(description="Path to region of interest file prefix by reality data id",
                                              alias="regionOfInterest",
                                              default=None)
    preset: Optional[str] = Field(default=None, description="Path to preset")


class ReferenceModelType(Enum):
    ORTHOPHOTO = "Orthophoto"
    COMPLETE = "Complete"


class TilingMode(Enum):
    NO_TILING = "NoTiling"
    REGULAR_PLANAR_GRID = "RegularPlanarGrid"
    REGULAR_VOLUMETRIC_GRID = "RegularVolumetricGrid"
    ADAPTIVE = "Adaptive"


class GeometricPrecision(Enum):
    MEDIUM = "Medium"
    HIGH = "High"
    EXTRA = "Extra"
    ULTRA = "Ultra"


class TilingPairSelection(Enum):
    GENERIC = "Generic"
    STRUCTURED_AERIAL = "StructuredAerial"
    REGION_OF_INTEREST = "RegionOfInterest"


class PhotoUsedForGeometry(Enum):
    EXCLUDE_THERMAL = "ExcludeThermal"
    INCLUDE_THERMAL = "IncludeThermal"
    NO = "None"


class HoleFilling(Enum):
    SMALL_HOLES = "SmallHoles"
    ALL_HOLES = "AllHoles"


class Simplification(Enum):
    STANDARD = "Standard"
    PLANAR_RELATIVE = "PlanarRelative"
    PLANAR_ABSOLUTE = "PlanarAbsolute"


class ColorCorrection(Enum):
    NO = "None"
    STANDARD = "Standard"
    STANDARD_WITH_THERMAL = "StandardWithThermal"
    BLOCK_WISE = "BlockWise"
    BLOCK_WISE_WITH_THERMAL = "BlockWiseWithThermal"


class UntexturedRepresentation(Enum):
    INPAINTING_COMPLETION = "InpaintingCompletion"
    UNIFORM_COLOR = "UniformColor"


class PointCloudColorSource(Enum):
    NO = "None"
    COLOR = "Color"
    INTENSITY = "Intensity"
    SCALED_INTENSITY = "ScaledIntensity"


class TextureSource(Enum):
    PHOTOS_FIRST = "PhotosFirst"
    POINT_CLOUDS_FIRST = "PointCloudsFirst"
    SMART = "Smart"


class ReferenceModelOptions(BaseModel):
    ref_model_type: Optional[ReferenceModelType] = Field(description="Reference Model Type", alias="referenceModelType",
                                                         default=None)
    tiling_mode: Optional[TilingMode] = Field(description="Tiling Mode", alias="tilingMode", default=None)
    tiling_value: Optional[float] = Field(description="Tiling Value", alias="tilingValue", ge=0, default=None)
    tiling_origin: Optional[Point3d] = Field(description="Tiling origin", alias="tilingOrigin", default=None)
    discard_empty_tiles: Optional[bool] = Field(description="Discard emtpy tiles", alias="discardEmptyTiles",
                                                default=None)
    srs: Optional[str] = Field(description="Spatial Reference System", default=None)
    geometric_precision: Optional[GeometricPrecision] = Field(description="Geometric precision",
                                                              alias="geometricPrecision", default=None)
    pair_selection: Optional[TilingPairSelection] = Field(description="Pair selection", alias="pairSelection",
                                                          default=None)
    photo_used_for_geometry: Optional[PhotoUsedForGeometry] = Field(description="Photo used for geometry",
                                                                    alias="photoUsedForGeometry", default=None)
    hole_filling: Optional[HoleFilling] = Field(description="Hole Filling", alias="holeFilling", default=None)
    simplification: Optional[Simplification] = Field(description="Simplification", default=None)
    planar_simplification_tolerance: Optional[float] = Field(description="Planar simplification tolerance",
                                                             alias="planarSimplificationTolerance", default=None)
    point_cloud_color_source: Optional[PointCloudColorSource] = Field(description="Point cloud color source",
                                                                      alias="pointCloudColorSource", default=None)
    color_correction: Optional[ColorCorrection] = Field(description="Color correction", alias="colorCorrection",
                                                        default=None)
    untextured_representation: Optional[UntexturedRepresentation] = Field(description="Untextured representation",
                                                                          alias="untexturedRepresentation",
                                                                          default=None)
    untextured_color: Optional[constr(pattern=r'^#[a-fA-F0-9]{6}$')] = Field(description="Untextured color",
                                                                             alias="untexturedColor", default=None)
    texture_source: Optional[TextureSource] = Field(description="Texture source", alias="textureSource", default=None)
    ortho_resolution: Optional[float] = Field(description="Ortho resolution", alias="orthoResolution", default=None)
    geometry_resolution_limit: Optional[float] = Field(description="Geometry resolution limit",
                                                       alias="geometryResolutionLimit", default=None)
    texture_resolution_limit: Optional[float] = Field(description="Texture resolution limit",
                                                      alias="textureResolutionLimit", default=None)


class ReferenceModel(BaseModel):
    reference_model_path: str = Field(description="Reality data id of reference model", alias="referenceModelPath")


class TilingOutputs(BaseModel):
    reference_model: ReferenceModel = Field(description="Reference model", alias="referenceModel")


class TilingOutputsCreate(BaseModel):
    REFERENCE_MODEL = "referenceModel"


class TilingSpecifications(BaseModel):
    inputs: TilingInputs = Field(description="Inputs")
    outputs: TilingOutputs = Field(description="Outputs")
    options: ReferenceModelOptions = Field(description="Options")


class TilingSpecificationsCreate(BaseModel):
    inputs: TilingInputs = Field(description="Inputs")
    outputs: list[TilingOutputsCreate] = Field(description="Outputs")
    options: ReferenceModelOptions = Field(description="Options")


class Coords2d(BaseModel):
    x: float = Field(description="X coordinate")
    y: float = Field(description="Y coordinate")


class Polygon2DWithHoles(BaseModel):
    outsideBounds: list[Coords2d] = Field(description="Outside bounds of the polygon")
    holes: Optional[list[list[Coords2d]]] = Field(default=None, description="List of holes boundaries if any")


class RegionOfInterest(BaseModel):
    srs: str = Field(description="Definition of the Region of Interest Coordinate System")
    polygons: list[Polygon2DWithHoles] = Field(description="List of polygons")
    altitude_min: float = Field(description="Minimum altitude", alias="altitudeMin")
    altitude_max: float = Field(description="Maximum altitude", alias="altitudeMax")


class LayoutTile(BaseModel):
    name: str = Field(description="Tile name")
    box_tight: BoundingBox = Field(description="Tight box encompassing the tile", alias="boxTight")
    box_overlapping: BoundingBox = Field(description="Overlapping box encompassing the tile and a bit of its neighbors",
                                         alias="boxOverlapping")
    memory_usage: float = Field(description="Memory usage of the tile", alias="memoryUsage", ge=0)


class Layout(BaseModel):
    tiles: list[LayoutTile] = Field(description="List of tiles in the layout")
    enu_definition: str = Field(description="Definition of the Internal Coordinate System", alias="enuDefinition")
    srs_definition: str = Field(description="Definition of the Layout Coordinate System", alias="srsDefinition")
    roi: RegionOfInterest = Field(description="Region of interest of the layout")

File: typescript_sdk/src/reality_capture/specifications/tiling.ts
import { z } from "zod";
import { BoundingBox, Point3d, RegionOfInterest } from "./geometry";

// Enums
export enum ModelingReferenceType {
  ORTHOPHOTO = "Orthophoto",
  COMPLETE = "Complete",
}

export enum TilingMode {
  NO_TILING = "NoTiling",
  REGULAR_PLANAR_GRID = "RegularPlanarGrid",
  REGULAR_VOLUMETRIC_GRID = "RegularVolumetricGrid",
  ADAPTIVE = "Adaptive",
}

export enum GeometricPrecision {
  MEDIUM = "Medium",
  HIGH = "High",
  EXTRA = "Extra",
  ULTRA = "Ultra",
}

export enum TilingPairSelection {
  GENERIC = "Generic",
  STRUCTURED_AERIAL = "StructuredAerial",
  REGION_OF_INTEREST = "RegionOfInterest",
}

export enum PhotoUsedForGeometry {
  EXCLUDE_THERMAL = "ExcludeThermal",
  INCLUDE_THERMAL = "IncludeThermal",
  NO = "None",
}

export enum HoleFilling {
  SMALL_HOLES = "SmallHoles",
  ALL_HOLES = "AllHoles",
}

export enum Simplification {
  STANDARD = "Standard",
  PLANAR_RELATIVE = "PlanarRelative",
  PLANAR_ABSOLUTE = "PlanarAbsolute",
}

export enum ColorCorrection {
  NO = "None",
  STANDARD = "Standard",
  STANDARD_WITH_THERMAL = "StandardWithThermal",
  BLOCK_WISE = "BlockWise",
  BLOCK_WISE_WITH_THERMAL = "BlockWiseWithThermal",
}

export enum UntexturedRepresentation {
  INPAINTING_COMPLETION = "InpaintingCompletion",
  UNIFORM_COLOR = "UniformColor",
}

export enum PointCloudColorSource {
  NO = "None",
  COLOR = "Color",
  INTENSITY = "Intensity",
  SCALED_INTENSITY = "ScaledIntensity",
}

export enum TextureSource {
  PHOTOS_FIRST = "PhotosFirst",
  POINT_CLOUDS_FIRST = "PointCloudsFirst",
  SMART = "Smart",
}

export enum TilingOutputsCreate {
  MODELING_REFERENCE = "modelingReference",
}

// Zod Schemas

export const TilingInputsSchema = z.object({
  scene: z.string().describe("Reality data id of ContextScene to process"),
  regionOfInterest: z
    .string()
    .regex(/^bkt:.+/)
    .describe("Path in the bucket to region of interest file")
    .optional(),
  presets: z.array(z.string()).describe("List of paths to preset").optional(),
});

export type TilingInputs = z.infer<typeof TilingInputsSchema>;

export const TilingOptionsSchema = z.object({
  modelingReferenceType: z.nativeEnum(ModelingReferenceType).describe("Modeling Reference Type").optional(),
  tilingMode: z.nativeEnum(TilingMode).describe("Tiling Mode").optional(),
  tilingValue: z.number().describe("Tiling Value").gte(0).optional(),
  tilingOrigin: Point3d.describe("Tiling origin").optional(),
  discardEmptyTiles: z.boolean().describe("Discard emtpy tiles").optional(),
  crs: z.string().describe("Coordinate Reference System").optional(),
  geometricPrecision: z.nativeEnum(GeometricPrecision).describe("Geometric precision").optional(),
  pairSelection: z.nativeEnum(TilingPairSelection).describe("Pair selection").optional(),
  photoUsedForGeometry: z.nativeEnum(PhotoUsedForGeometry).describe("Photo used for geometry").optional(),
  holeFilling: z.nativeEnum(HoleFilling).describe("Hole Filling").optional(),
  simplification: z.nativeEnum(Simplification).describe("Simplification").optional(),
  planarSimplificationTolerance: z.number().describe("Planar simplification tolerance").optional(),
  pointCloudColorSource: z.nativeEnum(PointCloudColorSource).describe("Point cloud color source").optional(),
  colorCorrection: z.nativeEnum(ColorCorrection).describe("Color correction").optional(),
  untexturedRepresentation: z.nativeEnum(UntexturedRepresentation).describe("Untextured representation").optional(),
  untexturedColor: z
    .string()
    .regex(/^#[a-fA-F0-9]{6}$/)
    .describe("Untextured color")
    .optional(),
  textureSource: z.nativeEnum(TextureSource).describe("Texture source").optional(),
  orthoResolution: z.number().describe("Ortho resolution").optional(),
  geometryResolutionLimit: z.number().describe("Geometry resolution limit").optional(),
  textureResolutionLimit: z.number().describe("Texture resolution limit").optional(),
});
export type TilingOptions = z.infer<typeof TilingOptionsSchema>;

export const ModelingReferenceSchema = z.object({
  location: z.string().describe("Reality data id of modeling reference"),
});
export type ModelingReference = z.infer<typeof ModelingReferenceSchema>;

export const TilingOutputsSchema = z.object({
  modelingReference: ModelingReferenceSchema.describe("Modeling reference"),
});
export type TilingOutputs = z.infer<typeof TilingOutputsSchema>;

export const TilingSpecificationsSchema = z.object({
  inputs: TilingInputsSchema.describe("Inputs"),
  outputs: TilingOutputsSchema.describe("Outputs"),
  options: TilingOptionsSchema.optional().describe("Options"),
});
export type TilingSpecifications = z.infer<typeof TilingSpecificationsSchema>;

export const TilingSpecificationsCreateSchema = z.object({
  inputs: TilingInputsSchema.describe("Inputs"),
  outputs: z.array(z.nativeEnum(TilingOutputsCreate)).describe("Outputs"),
  options: TilingOptionsSchema.optional().describe("Options"),
});
export type TilingSpecificationsCreate = z.infer<typeof TilingSpecificationsCreateSchema>;

export const TilingCostSchema = z.object({
  gpix: z.number().describe("Number of GigaPixels in the overall inputs, after applying downsampling.").gte(0),
  mpoints: z.number().describe("Number of MegaPoints in the overall inputs.").gte(0),
});
export type TilingCost = z.infer<typeof TilingCostSchema>;

export const LayoutTileSchema = z.object({
  name: z.string().describe("Tile name"),
  boxTight: BoundingBox.describe("Tight box encompassing the tile"),
  boxOverlapping: BoundingBox.describe(
    "Overlapping box encompassing the tile and a bit of its neighbors"
  ),
  memoryUsage: z.number().describe("Memory usage of the tile").gte(0),
});
export type LayoutTile = z.infer<typeof LayoutTileSchema>;

export const LayoutSchema = z.object({
  tiles: z.array(LayoutTileSchema).describe("List of tiles in the layout"),
  enuDefinition: z.string().describe("Definition of the Internal Coordinate System"),
  crsDefinition: z.string().describe("Definition of the Layout Coordinate System"),
  roi: RegionOfInterest.describe("Region of interest of the layout"),
});
export type Layout = z.infer<typeof LayoutSchema>;